### ISSUE TITLE
Make select() understand OS/2 pipes created with pipe()

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/libcx#34]
Many (all?) Unix applications use handles created with `pipe()` for stdio redirection to capture output of other programs (usually forked with `fork()`). On OS/2 we have to use `socketpair` instead of `pipe` because handles returned by `pipe` are native OS/2 pipe handles and they are not compatible with the current kLIBC implementation of `select()` which is used by these Unix applications to read captured output and such.

`socketpair()` is great but it only works if applications whose output is redirected are also kLIBC-based applications. kLIBC detects that the given fd is a socket and calls appropriate OS/2 TCP/IP socket APIs (send/recv and alike). However, if the application is not kLIBC-based (e.g. it's a native OS/2 app or an app using the IBM VAC runtime), it knows nothing about file descriptors being sockets, they are not inherited and they cannot be written to. As a result, the calling application gets no output from the callee.

A real world example is our kLIBC-based python — when it is asked to capture output of a non-kLIBC-based program, it captures nothing. See e.g. http://trac.netlabs.org/rpm/ticket/136.

**Comment by dmik** [comment]
The only ultimate solution here is to use `pipe` instead of `socketpair` and teach kLIBC `select()` support native OS/2 pipes. This will make everybody happy as handlers returned by `pipe` may be passed directly to DosWrite/DosRead etc.

Note that kLIBC is based on the EMX runtime which used to support OS/2 pipes as well as normal OS/2 file handles in addition to TCP/IP sockets in its `select` implementation and this implementation is even still present in the kLIBC sources (unused and unavailable from the outer world). The reason why Knut replaced it with a new `select()` that only supports TCP/IP sockets is unclear to me. Perhaps, this has something to do with `fork()` emulation  or with performance issues.

A way to go here is to create a new `select` implementation in LIBCx that will deal with at least OS/2 pipes. We may look at the old EMX implementation to see how they did it back then.

**Comment by dmik** [comment]
BTW, I've just got another idea on how to make non-kLIBC-based apps work with `socketpair` (which doesn't require `select` support for OS/2 pipes) — create a kLIBC-based EXE wrapper that will start a non-kLIBC-based program, capture its output with native OS/2 pipes and write it to its own stdout (which can be attached to a socket so redirection set up by the parent will work properly). Not directly related to this ticket but may be worth doing if one urgently needs to capture output of some non-kLIBC-based tool and doesn't want to wait for a better `select` implementation.

**Comment by ydario** [comment]
select() was supporting pipe handles in EMX, maybe you can grab some code.

**Comment by dmik** [comment]
@ydario yes, that's exactly what I write in comment 2 :) Do you by any chance know why Knut dropped it?

**Comment by ydario** [comment]
:-)
no idea.

**Comment by dryeo** [comment]
On 03/22/17 10:26 am, Dmitriy Kuminov wrote:
> 2 :) Do you by any chance know why Knut dropped it?

Reviewing some old mailing lists, seems that EMX file handles weren't 
quite compatible with kLIBC ones making supporting EMX select() 
non-trivial. Someone else promised to add the support and never came 
through and Knut did talk about fixing it but apparently never got 
around to it. There was quite a bit of internal EMX stuff removed when 
kLIBC was created.
Dve

**Comment by psmedley** [comment]
AFAIR Brian Smith (aka Nuke) had code to implement this but had a hard
drive crash or similar and lost the code.

Cheers,

Paul

**Comment by abwillis** [comment]
Yes, Nuke was the one that had implemented it, and because he had, Knut did not as he was going to use Brian's but then Brian lost the code.

**Comment by SilvanScherrer** [comment]
related to #84
